prov/util/src/util_mr.c

```diff
@@ -0,0 +1,223 @@
+            return 0;
+    }
+
+    return -FI_EACCES;
+}
+
+int ofi_mr_insert(struct util_mr * in_mr_h, const struct fi_mr_attr *in_attr,
+                                uint64_t * out_key, void * in_prov_mr)
+{
+    util_mr_item_t * item;
+
+    if (!in_attr || in_attr->iov_count <= 0 || !in_prov_mr) {
+        return -FI_EINVAL;
+    }
+
+    item = create_mr_attr_copy(in_attr, in_prov_mr);
+
+    /* Scalable MR handling: use requested key and offset */
+    if(in_mr_h->mr_type == FI_MR_SCALABLE) {
+        item->offset = (uintptr_t) in_attr->mr_iov[0].iov_base + in_attr->offset;
+        /* verify key doesn't already exist */
+        if(rbtFind(in_mr_h->map_handle, &item->requested_key))
+                return -FI_EINVAL;
+    } else {
+        item->requested_key = get_mr_key(in_mr_h);
+        item->offset = (uintptr_t) in_attr->mr_iov[0].iov_base;
+    }
+
+    rbtInsert(in_mr_h->map_handle, &item->requested_key, item);
+    *out_key = item->requested_key;
+
+    return 0;
+}
+
+/* io_addr is address of buff (&buf) */
+int ofi_mr_retrieve(struct util_mr * in_mr_h, ssize_t in_len,
+                                uintptr_t *io_addr, uint64_t in_key,
+                                uint64_t in_access, void **out_prov_mr)
+{
+    /*grab info */
+    util_mr_itr itr;
+    util_mr_item_t * item;
+    int err = 0;
+
+    assert(io_addr);
+
+    itr = rbtFind(in_mr_h->map_handle, &in_key);
+
+    if(!itr)
+        return -FI_ENOKEY;
+
+    rbtKeyValue(in_mr_h->map_handle, itr, (void **)&in_key,
+                                (void **) &item);
+
+    /*return providers MR struct */
+    if(!item)
+        return -FI_EKEYREJECTED;
+
+    (*out_prov_mr) = item->context;
+
+    /*offset for scalable */
+    if(in_mr_h->mr_type == FI_MR_SCALABLE)
+        *io_addr = (*io_addr) + item->offset;
+
+    err = verify_addr(item, in_access, *io_addr, in_len);
+    if(err)
+       return err;
+
+
+    return 0;
+}
+
+int ofi_mr_erase(struct util_mr * in_mr_h, uint64_t in_key)
+{
+    util_mr_itr itr;
+    util_mr_item_t * item;
+
+    if(!in_mr_h)
+        return -FI_EINVAL;
+
+    itr = rbtFind(in_mr_h->map_handle, &in_key);
+
+    if(!itr)
+        return -FI_ENOKEY;
+
+    /*release memory */
+    rbtKeyValue(in_mr_h->map_handle, itr, (void **)&in_key,
+                                (void **) &item);
+
+    assert(item);
+
+    free((void *)item->mr_iov);
+    free(item);
+
+    rbtErase(in_mr_h->map_handle, itr);
+
+    return 0;
+}
+
+/*assumes uint64_t keys */
+static int compare_mr_keys(void *key1, void *key2)
+{
+    uint64_t k1 = *((uint64_t *) key1);
+    uint64_t k2 = *((uint64_t *) key2);
+    return (k1 < k2) ?  -1 : (k1 > k2);
+}
+
+
+struct util_mr * ofi_mr_init(enum fi_mr_mode mode)
+{
+    struct util_mr * new_mr = malloc(sizeof(struct util_mr));
+
+    assert((mode == FI_MR_SCALABLE) || (mode == FI_MR_BASIC));
+
+    new_mr->mr_type = mode;
+
+    new_mr->map_handle = rbtNew(compare_mr_keys);
+
+    assert(new_mr->map_handle);
```

Comment:
better check the value and return error here. assert is more suitable for errors that is not supposed to happen, and assert is nop for non-debug build.